pallene/translator.lua

```diff
@@ -0,0 +1,81 @@
+local translator = {}
+
+local function translate_expression(expression)
+    if expression._tag == "ast.Exp.Integer" then
+        io.write(expression.value)
+    end
+end
+
+local function translate_tl_var(node)
+    io.write('local ')
+    for _, declaration in pairs(node.decls) do
+        io.write(declaration.name)
+    end
+
+    if #node.values > 0 then
+        io.write(' = ')
+        for _, value in pairs(node.values) do
+            translate_expression(value)
+        end
+    end
+
+    io.write('\n')
+end
+
+local last_index = 1
+
+local function add_previous(input, partials, stop_index)
+    local partial = input:sub(last_index, stop_index)
+    table.insert(partials, partial)
+
+    -- Do not update the last index if this is the last call to add_previous.
+    if stop_index then
+        last_index = stop_index + 1
+    end
+end
+
+-- TODO: Added test case for newlines, space, comments, and tabs.
+
+local function add_whitespace(input, partials, start_index, stop_index)
+    add_previous(input, partials, start_index - 1)
+
+    local p = start_index
+    local q = start_index
+    while q <= stop_index do
+        if input:sub(q, q) == "\n" then
+            local partial = string.rep(" ", q - p - 1)
+            table.insert(partials, partial)
+            table.insert(partials, "\n")
+            p = q + 2
+        end
+        q = q + 1
+    end
+    local final_partial = string.rep(" ", q - p)
+    table.insert(partials, final_partial)
```

Comment:
What do you think about using `string.gsub` here? Replace any non-whitespace characters with a space `string.replace(str, "%S", " ").

By the way, this reminds me that we probably should also be careful about preserving tabs, for proper indentation, and `\r`, for people who are programming on Windows. Could you please add a test case that uses hard tabs?

Reply:
I actually have a list of things to test, tabs are included. But then I am waiting to implement the translator for the top-level variables completely before writing them.